barretenberg/cpp/src/barretenberg/vm2/tracegen/note_hash_tree_check_trace.cpp

```diff
@@ -19,72 +20,116 @@ void NoteHashTreeCheckTraceBuilder::process(
 {
     using C = Column;
 
-    uint32_t row = 0;
-    for (const auto& event : events) {
-        bool write = event.append_data.has_value();
-        bool discard = false; // TODO: Reconstruct discard
-
-        FF note_hash = event.note_hash;
-        FF siloed_note_hash = event.note_hash;
-        FF unique_note_hash = event.note_hash;
-
-        bool should_silo = false;
-        AztecAddress address = 0;
-        FF nonce = 0;
-
-        bool should_unique = false;
-        uint64_t note_hash_counter = 0;
-        FF next_root = 0;
-        FF first_nullifier = 0;
-        FF prev_leaf_value = 0;
-
-        if (write) {
-            simulation::NoteHashAppendData append_data = event.append_data.value();
-            should_silo = append_data.siloing_data.has_value();
-            should_unique = append_data.uniqueness_data.has_value();
-            if (should_silo) {
-                siloed_note_hash = append_data.siloing_data->siloed_note_hash;
-                address = append_data.siloing_data->address;
+    // Maps the index of the checkpoint to the index it was reverted in.
+    std::unordered_map<size_t, size_t> reverted_in;
+
+    std::vector<size_t> checkpoint_stack;
```

Comment:
Prefer std::stack if this is truly used as a stack. Probably doesn't make any difference though.